### PR TITLE
don't allow to create a federated share if source and target are the same

### LINKED ISSUE #20296: [8.1] Mount fed share as same user, move source into received folder, must be forbidden
### Steps

1. Create a folder "abc"
1. Share "abc" with link and allow editing
1. Open link
1. Click "add to ownCloud"
1. Accept the share as the same user who shared it, it will appear as "abc (2)"
1. Move "abc" (the source folder) into "abc (2)"

### Expected result
Moving not allowed (?)

### Actual result
Moving hangs. Refresh the page. "abc" is simultaneously in the root and inside "abc (2)"

### Versions
Observed on 8.1.4.

On 8.2.0 this operation doesn't work and you get an error that "abc (2)" is locked, which makes sense as it's the same folder.

```
MariaDB [owncloud]> select storage,fileid,path,size from oc_filecache;
+---------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+
| storage | fileid | path                                                                                                                                                                          | size  |
+---------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+
|       1 |      1 |                                                                                                                                                                               |   163 |
|       1 |      2 | files                                                                                                                                                                         |   163 |
|       1 |      3 | files/welcome.txt                                                                                                                                                             |   163 |
|       1 |      4 | thumbnails                                                                                                                                                                    | 49382 |
|       1 |      5 | thumbnails/3                                                                                                                                                                  | 49382 |
|       1 |      6 | thumbnails/3/2048-2048-max.png                                                                                                                                                | 45453 |
|       1 |      7 | thumbnails/3/61-61.png                                                                                                                                                        |  3929 |
|       1 |      8 | files/abc                                                                                                                                                                     |     0 |
|       3 |      9 |                                                                                                                                                                               |     0 |
|       1 |     10 | files/abc/abc                                                                                                                                                                 |     0 |
|       1 |     11 | files/abc/abc/abc                                                                                                                                                             |     0 |
|       1 |     12 | files/abc/abc/abc/abc                                                                                                                                                         |     0 |
|       1 |     13 | files/abc/abc/abc/abc/abc                                                                                                                                                     |     0 |
|       1 |     14 | files/abc/abc/abc/abc/abc/abc                                                                                                                                                 |     0 |
|       1 |     15 | files/abc/abc/abc/abc/abc/abc/abc                                                                                                                                             |     0 |
|       1 |     16 | files/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                                         |     0 |
|       1 |     17 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                                     |     0 |
|       1 |     18 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                                 |     0 |
|       1 |     19 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                             |     0 |
|       1 |     20 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                         |     0 |
|       1 |     21 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                     |     0 |
|       1 |     22 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                 |     0 |
|       1 |     23 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                             |     0 |
|       1 |     24 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                         |     0 |
|       1 |     25 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                     |     0 |
|       1 |     26 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                 |     0 |
|       1 |     27 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                             |     0 |
|       1 |     28 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                         |     0 |
|       1 |     29 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                     |     0 |
|       1 |     30 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                 |     0 |
|       1 |     31 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                             |     0 |
|       1 |     32 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                         |     0 |
|       1 |     33 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                     |     0 |
|       1 |     34 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                 |     0 |
|       1 |     35 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                             |     0 |
|       1 |     36 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                         |     0 |
|       1 |     37 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                     |     0 |
|       1 |     38 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                 |     0 |
|       1 |     39 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                             |     0 |
|       1 |     40 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                         |     0 |
|       1 |     41 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                     |     0 |
|       1 |     42 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                 |     0 |
|       1 |     43 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                             |     0 |
|       1 |     44 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                         |     0 |
|       1 |     45 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                     |     0 |
|       1 |     46 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                 |     0 |
|       1 |     47 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc             |     0 |
|       1 |     48 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc         |     0 |
|       1 |     49 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc     |     0 |
|       1 |     50 | files/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc |     0 |
|       3 |     51 | abc                                                                                                                                                                           |     0 |
|       3 |     52 | abc/abc                                                                                                                                                                       |     0 |
|       3 |     53 | abc/abc/abc                                                                                                                                                                   |     0 |
|       3 |     54 | abc/abc/abc/abc                                                                                                                                                               |     0 |
|       3 |     55 | abc/abc/abc/abc/abc                                                                                                                                                           |     0 |
|       3 |     56 | abc/abc/abc/abc/abc/abc                                                                                                                                                       |     0 |
|       3 |     57 | abc/abc/abc/abc/abc/abc/abc                                                                                                                                                   |     0 |
|       3 |     58 | abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                                               |     0 |
|       3 |     59 | abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                                           |     0 |
|       3 |     60 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                                       |     0 |
|       3 |     61 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                                   |     0 |
|       3 |     62 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                               |     0 |
|       3 |     63 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                           |     0 |
|       3 |     64 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                       |     0 |
|       3 |     65 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                                   |     0 |
|       3 |     66 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                               |     0 |
|       3 |     67 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                           |     0 |
|       3 |     68 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                       |     0 |
|       3 |     69 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                                   |     0 |
|       3 |     70 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                               |     0 |
|       3 |     71 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                           |     0 |
|       3 |     72 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                       |     0 |
|       3 |     73 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                                   |     0 |
|       3 |     74 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                               |     0 |
|       3 |     75 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                           |     0 |
|       3 |     76 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                       |     0 |
|       3 |     77 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                                   |     0 |
|       3 |     78 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                               |     0 |
|       3 |     79 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                           |     0 |
|       3 |     80 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                       |     0 |
|       3 |     81 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                                   |     0 |
|       3 |     82 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                               |     0 |
|       3 |     83 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                           |     0 |
|       3 |     84 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                       |     0 |
|       3 |     85 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                                   |     0 |
|       3 |     86 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                               |     0 |
|       3 |     87 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                           |     0 |
|       3 |     88 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                       |     0 |
|       3 |     89 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc                   |     0 |
|       3 |     90 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc               |     0 |
|       3 |     91 | abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc/abc           |     0 |
+---------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+
```

Not sure if this case can even be detected properly, especially if the users have a chain of federated shares!

### PR DESCRIPTION
**Pull request by schiessle**
fix #20296 

@rullzer I added the fix to the old code but we should make sure that it also went into the new sharing. Afaik there is no sharing 2.0 code yet where I could add it right away, right?

@karlitschek should be backported to 8.2, 8.1 and probably also to 8.0 for the maintenance releases

### CODE REVIEW

**Comment by rullzer** [comment]
There is indeed no new federated sharing code yet

File: apps/files_sharing/ajax/external.php [hunk 1]
```diff
@@ -49,6 +49,14 @@
 	exit();
 }
 
+$target = \OC\Share\Share::removeProtocolFromUrl(\OC::$server->getURLGenerator()->getAbsoluteURL('/'));
+$source = \OC\Share\Share::removeProtocolFromUrl($remote);
```
Comment (PVince81) [inline apps/files_sharing/ajax/external.php]: Would it make sense to have a unit test for this ? Maybe move it into a utility function that can be called from a test and use a data provider to pass several URL formats ? (with an without trailing slash).

**Comment by nickvergessen** [comment]
Uhhh this is quite bad for testing.

I see the idea behind preventing that, but it will be a mess to test remote sharing locally, with integration tests and with smashbox then.

I'd really like to see this avoidable or at least configurable (debug?)

**Comment by PVince81** [comment]
Wait, does this block sharing with any users locally ? I think it should only be blocked if a user is sharing with themselves. Sharing with other users on the same instance should be allowed for now, until we have https://github.com/owncloud/core/issues/9090

**Comment by schiesbn** [comment]
I don't think it makes sense to allow a remote share between the same server (smashbox tests excluded). The user should create a local share in this case.
@nickvergessen I thought that smashbox also set-up two different servers to test server-to-server shares

> Wait, does this block sharing with any users locally ? I think it should only be blocked if a user is sharing with themselves. Sharing with other users on the same instance should be allowed for now, until we have #9090

I see it the other way around. As long as we don't have #9090 we should block it completely, because it doesn't make sense at all. With #9090 the blocking will be bypassed because we will transform the share from a local share to a remote share at a earlier stage.

**Comment by nickvergessen** [comment]
> @nickvergessen I thought that smashbox also set-up two different servers to test server-to-server shares

No, it fires up one docker image (on jenkins)

If we would do that, we would need to git clone twice, double the server configuration, double half the smashbox methods to allow specifying on which server the action should be run, etc. This will become really messy.

File: apps/files_sharing/ajax/external.php [hunk 2]
```diff
@@ -49,6 +49,14 @@
...
+if (rtrim($source, '/') === rtrim($target, '/')) {
```
Comment (nickvergessen) [inline apps/files_sharing/ajax/external.php]: if we do this, we should compare against trusted domains?

**Comment by PVince81** [comment]
@schiesbn I'd agree with the blocking if federation was a new features. Unfortunately there are already people out there using this feature to mount remote links.

To avoid breaking the "feature" for existing users by changing the behavior, I suggest we only block for the current user. If you still think we should block completely, then we could introduce this behavior change in 9.0. (not wanting behavior changes in minor releases)

How does that sound ?

**Comment by PVince81** [comment]
@MorrisJobke what do you think ? (based on known real life use cases)

**Comment by schiesbn** [comment]
@nickvergessen I see the point, but I'm still not convinced. This would add quite some additional complexity. For example for some user back-ends user names are case sensitive and for some they are not. We would take all this into account if we want to compare the user name properly. I think this is way to complex and error-prone. If we take one step back, we need to admit that allowing remote shares between the same servers was a bug from the beginning. That's just not what they are for and we shouldn't allow it.

**Comment by schiesbn** [comment]
> To avoid breaking the "feature" for existing users by changing the behavior, I suggest we only block for the current user. If you still think we should block completely, then we could introduce this behavior change in 9.0. (not wanting behavior changes in minor releases)

I don't think it breaks a "feature" it was/is a bug. Some people might have utilized it, right. But we should still fix it. Existing shares will still work. But we shouldn't keep a bug just because some people utilized it in the past.

**Comment by PVince81** [comment]
Your argument about being unable to properly compare user names is the biggest issue now.
Can we ask the UserManager to return the user matching that specific user and see if it's the same on the local instance ? If UserManager->get('USER1') returns 'user1' then we know it's the same. If it doesn't, then it's not the same. Basically, make the user name comparison be the job of the UserManager which knows better whether user names are case sensitive or not. Would that solve the user name issue ?

How are we dealing with this issue regarding user names in other parts of federated sharing ?

**Comment by schiesbn** [comment]
@nickvergessen I don't want to add a specific switch for it but you can be quite sure that if such a switch exists people will find and use it. How about to skip the test if the debug mode is enabled? Still a hack but at least it would make some sense to allow it for debugging purpose.

**Comment by schiesbn** [comment]
@PVince81 Homestly I would really like to avoid to maintain a complex fix just to allow people to continue to explore a bug for a few more weeks/months (depending how fast they upgrade) instead of simply fixing it.

**Comment by PVince81** [comment]
@cmonteroluque @karlitschek @MTRichards this PR will block federated shares sent the the same ownCloud instance, and also prevent mounting link shares from the same instance using "Add to ownCloud". Do we agree with this change of behavior on the basis that it was originally a bug ?

Please have a look at the discussion in this PR to understand the reasoning. Thanks.

**Comment by PVince81** [comment]
(the original intent of this PR was to prevent recursively sharing with yourself on the same instance and potentially creating infinite recursion loops, see https://github.com/owncloud/core/issues/20296)

**Comment by MorrisJobke** [comment]
> If we take one step back, we need to admit that allowing remote shares between the same servers was a bug from the beginning.

Woot? This is the coolest feature ever! Don't kill it.

Why is it hard to detect if this is from the same instance? Just lookup the secret token in the local database and compare the base of the URL -> if this is the case maybe also convert it into a local share. see https://github.com/owncloud/core/issues/9090

**Comment by MorrisJobke** [comment]
> @cmonteroluque @karlitschek @MTRichards this PR will block federated shares sent the the same ownCloud instance, and also prevent mounting link shares from the same instance using "Add to ownCloud". Do we agree with this change of behavior on the basis that it was originally a bug ?

I don't think so :-1: from me then.

**Comment by schiesbn** [comment]
> Woot? This is the coolest feature ever! Don't kill it.

Why is this a feature? Federated sharing exists to share between multiple ownClouds. If we allow it within the same ownCloud we only create problem. We have user and group sharing for it. Don't make it more complicated and complex as needed!

**Comment by PVince81** [comment]
@MorrisJobke see https://github.com/owncloud/core/pull/21067#issuecomment-163194429
The complexity is about matching user names because of different user backends.

I thought using the user manager could work https://github.com/owncloud/core/pull/21067#issuecomment-163196995 but not sure whether it's just a one line fix or whether many cases need to taken into consideration. Might be worth a try ? That approach will be needed anyway for https://github.com/owncloud/core/issues/9090 so if it doesn't work at all, then it won't work for https://github.com/owncloud/core/issues/9090 either.

**Comment by PVince81** [comment]
Also see this older discussion that might contain other arguments regarding the user backend difficulty: https://github.com/owncloud/core/pull/14218

File: tests/lib/share/helper.php [hunk 3]
```diff
@@ -121,4 +121,33 @@ public function dataTestSplitUserRemoteError() {
...
 	}
+
+	/**
+	 * @dataProvider dataTestCompareServerAddresses
+	 *
+	 * @param string $server1
+	 * @param string $server2
+	 * @param bool $expected
+	 */
+	public function testCompareServerAddresses($server1, $server2, $expected) {
```
Comment (PVince81) [inline tests/lib/share/helper.php]: Thanks!

**Comment by schiesbn** [comment]
> I thought using the user manager could work #21067 (comment) but not sure whether it's just a one line fix or whether many cases need to taken into consideration. Might be worth a try ? That approach will be needed anyway for #9090 so if it doesn't work at all, then it won't work for #9090 either.

Therefore we first have to decide if and how we want to fix #9090. Please don't mix to much stuff into it. The idea to convert a remote share to a internal share sounds convenient on one hand but on the other hand it also adds quite some magic. If we are both on the same ownCloud and I send you a link instead of a user share then I probably did this by purpose. Not sure if I really want that you are able to convert this to a user share. But let's this discuss here #9090.

As said, no matter what we decide in #9090 this doesn't affect this decision. All arguments are above, I don't want to repeat it. Just one more sentence: I just don't think we should make a bug a supported feature just because one or two user might have utilized the bug to generate a unintended behavior. Federated sharing was never meant to be a third API to generate local shares.

**Comment by schiesbn** [comment]
@nickvergessen Test_DB_QueryBuilder_QueryBuilderTest/testGetLastInsertId/ fails on pgsql, any idea: https://ci.owncloud.org/job/core-ci-linux/database=pgsql,label=SLAVE/15399/testReport/junit/%28root%29/Test_DB_QueryBuilder_QueryBuilderTest/testGetLastInsertId/

**Comment by karlitschek** [comment]
No strong opinions either way from me. I'm fine with killing this because this is not the real original usecase. But if there are real use-cases as @MorrisJobke said then maybe support it? 
Killing it would be the optimal solution because this is a weird fug (feature/bug)

**Comment by MorrisJobke** [comment]
> But if there are real use-cases as @MorrisJobke said then maybe support it?

I think there are a lot of use cases in the wild:

From a user perspective I usually don't care about who is on my instance (If you are not involved in the tech). I simply generate a link, send it via messenger/email/... to a friend and then can be used to mount it. I would really hate it as a user when this is only possible when both of us are on different instances. I often do that here too, because somebody just want to share some stuff with me and I want the freedom to hook it into the owncloud I want.

I think this would be a huge drawback once we deny mounting link shares from the same instance just because we are not really able to detect that there is a loop. 

And I also don't get why the username case is a problem? The share owner username is listed in oc_share (the outgoing share) and the current user is logged in. Where is the problem with the case and different backends, and ...

**Comment by schiesbn** [comment]
What you describe if what we discuss here #9090. If we want to allow it we should convert a remote share to a local share. We can discuss there if we want to implement it. But this shouldn't affect this discussion. (federated) sharing is already complex enough, don't make it even more complex by adding all possible features just because someone might want to use it. We have everything users might want: user, group, link and federated sharing. Let's keep the lines sharp and don't blur it. This makes it much easier to maintain and enhance the individual parts.

**Comment by MTRichards** [comment]
We really should not allow sharing on the same instance. 

It can create policy violations easily, where user A has one set of permissions, User B another on the same file. Then User A shares on the same instance to User B and voila - User B now has those new permissions. That, and it will duplicate disk space for the same files...and a bunch of other problems.

**Comment by MorrisJobke** [comment]
> That, and it will duplicate disk space for the same files...and a bunch of other problems.

How that?

**Comment by MorrisJobke** [comment]
> It can create policy violations easily, where user A has one set of permissions, User B another on the same file. Then User A shares on the same instance to User B and voila - User B now has those new permissions.

Also possible with a normal link share, download, upload, voila - you have the permissions

**Comment by MorrisJobke** [comment]
And you can't enforce reshaping permission via remote because you don't own the remote machine... once you grant read to a remote you lose the permissions for forbidden reshare. So this is not a valid argument.

**Comment by cmonteroluque** [comment]
@karlitschek, @MTRichards, and I just went through this. @schiesbn please fix only the "sharing with yourself" case. In exchange, we are also adding  #9090 to 9.0, so the case of federating with other local users will be covered by then. @MorrisJobke fyi

**Comment by rullzer** [comment]
I have to agree with @schiesbn. Users might have used it but it is damn near impossible to do this right. Also because #9090 is missing this will result in a significant hit in performance because you do a local webdav mount (Which is still orders of magnitude slower than a local share). I get the use case but this was never what it was intented/designed for. Thus it is very very suboptimal.

Solution for now to only block share with self is fine with me. But we should kill this feature ASAP.

I have objectsion agains #9090 in 9.0 but I'll voice them in that ticket.

**Comment by cmonteroluque** [comment]
OK, so I got a lot of pushback today, but I am still not sure that we should change the plan in 
https://github.com/owncloud/core/pull/21067#issuecomment-163387699 . I asked @schiesbn to not work on this until we have more conversations today. We will come up with a decision and stick with it

**Comment by PVince81** [comment]
Based on the last email discussion, the result is that local federated shares must be blocked only for the same user, for now. Backportability is still to be evaluated.

CC @DeepDiver1975 ^

**Comment by schiesbn** [comment]
Should work now as discussed... We block federated shares if sender and recipient are the same. Can someone please test this with LDAP users? I have some problems getting my LDAP running... @blizzz maybe? Thanks!

Steps to test:

With the share dialog:

- Login as a user and try to share a file/folder to the same user as a federated share ````<user>@<server>```` -> should return a error message if you try to share to the same user

public link share:

- Share a file/folder by link
- Open the share link
- click on add to your ownCloud and try to add the link to the same user who created the link share.

**Comment by PVince81** [comment]
@schiesbn https://github.com/owncloud/administration/tree/master/ldap-testing run `./start.sh`

**Comment by MTRichards** [comment]
Thank you.

**Comment by PVince81** [comment]
I justed tested with OpenLDAP, standard settings.
- [ ] If "zombie1" shares with "zombie1@localhost/owncloud" it goes through :no_entry_sign: FAIL
- [x] If "zombie1" shares with "b8261b06-36b2-1035-9a72-cd0b96131b42@localhost/owncloud", it is properly blocked

Note, I'm able to share with "zombie2" and the share goes through, so it seems that so far both names are supported for shares.

**Comment by PVince81** [comment]
It works for regular users but doesn't work properly for LDAP users, see above.

How about using `UserManager.get()` here https://github.com/owncloud/core/pull/21067/files#diff-a3370cea652fd59e87331d01e531b550R308 to get the user ? With a bit of luck maybe it will retrieve the same user for both inputs ? Then compare the `getUID()` to see whether it matches ?

@blizzz do you have some insights ?

File: lib/private/share/helper.php [hunk 4]
```diff
@@ -289,4 +289,27 @@ public static function splitUserRemote($id) {
 		$hint = $l->t('Invalid Federated Cloud ID');
 		throw new HintException('Invalid Fededrated Cloud ID', $hint);
 	}
+
+	/**
+	 * check if two federated cloud IDs refer to the same user
+	 *
+	 * @param string $user1
+	 * @oaram string $server1
```
Comment (nickvergessen) [inline lib/private/share/helper.php]: `@param`

**Comment by schiesbn** [comment]
got my ldap configured correctly and seems to work with this dirty hack: https://github.com/owncloud/core/pull/21067/files#diff-a3370cea652fd59e87331d01e531b550R308

**Comment by PVince81** [comment]
@schiesbn the new dirty hack also works for me, I ticked the checkboxes. However the link share mounting still lets it through for LDAP users.

**Comment by PVince81** [comment]
Just had a quick test locally with `UserManager->get()` and noticed that it only accepts one format, so not a viable solution.

**Comment by schiesbn** [comment]
> However the link share mounting still lets it through for LDAP users.

I can't reproduce it. How did you tested it? You need to go through all steps. Only after you click "add" on the dialog which asks you if you want to add the remote share, the check gets called and you should get a error message.

**Comment by PVince81** [comment]
Steps to reproduce:

1. Login as "zombie10" (LDAP user)
1. Create a folder "test"
1. Share "test" with link
1. Open the link
1. Click "Add to ownCloud", enter "http://localhost/owncloud"
1. The accept dialog pops up with the user's real name + "@localhost/owncloud" (not the user id), click "Accept"
1. Mount appears but shouldn't

**Comment by PVince81** [comment]
Looks like for some reason I'm getting the user's display name.
Before the hooks call:
```
// by property_get
$user1                           = (string[12]) 'Kissel, Elif';


// by property_get
$user2                           = (string[36]) 'b82c29a6-36b2-1035-9a79-cd0b96131b42';
```

After the call, still the same values:
```
// by property_get
$user1                           = (string[12]) 'Kissel, Elif';


// by property_get
$user2                           = (string[36]) 'b82c29a6-36b2-1035-9a79-cd0b96131b42';
```

**Comment by PVince81** [comment]
indeed, it seems that the display name is used when building the URL: `http://localhost/owncloud/index.php/apps/files#remote=http%3A%2F%2Flocalhost%2Fowncloud&token=s4XTDvegBXI4BJ7&owner=Kissel%2C%20Elif&name=test&protected=0`

**Comment by schiesbn** [comment]
Now it should also work with display names

**Comment by PVince81** [comment]
Tested, all work now :+1:

**Comment by PVince81** [comment]
Second reviewer @blizzz @rullzer ?

@karlitschek backport to 8.2, 8.1 and 8.0 ?

**Comment by karlitschek** [comment]
This needs to be backported. Thanks. :+1:

**Comment by schiesbn** [comment]
@PVince81 some jS unit tests fail... can you have a look? Thanks!

**Comment by PVince81** [comment]
Fixed JS unit test for the new field: https://github.com/owncloud/core/commit/1d37e7abfc8316be8f197eb250d04fbf80299a72

**Comment by PVince81** [comment]
stable8.2: https://github.com/owncloud/core/pull/21198
stable8.1: https://github.com/owncloud/core/pull/21199
stable8.0: https://github.com/owncloud/core/pull/21202

Experiencing some difficulties with 8.1 and 8.0 right now, stay tuned.

**Comment by rullzer** [comment]
Nice :+1: tested and works here. Code looks good!

**Pull request merged**